File: graph/arch/1584-min-cost-connect-points/golang/min-cost-connect-points-kruskal.go
package main

import "sort"

func kruskal(n int, xs []Edge) (t []Edge) {
	sort.Slice(xs, func(i, j int) bool { return xs[i].w < xs[j].w })
	u := NewUnionFind(n)
	for i := 0; i < len(xs) && len(t) < n-1; i++ {
		if u.Connect(xs[i].u, xs[i].v) {
			t = append(t, xs[i])
		}
	}
	return
}

func minCostConnectPointsKruskal(vs [][]int) (r int) {
	n := len(vs)
	xs := edges(vs)
	t := kruskal(n, xs)
	r = 0
	for _, x := range t {
		r = r + x.w
	}
	return
}

File: graph/arch/1584-min-cost-connect-points/py/min_cost_connect_points_kruskal.py
from typing import List
from union_find import UnionFind
from min_cost_connect_points import edges, Point, Edge


class SolutionKruskal:
  def minCostConnectPoints(self, pts: List[Point]) -> int:
    n = len(pts)
    xs = edges(pts)
    txs = kruskal(n, xs)
    r = sum(e[2] for e in txs)
    return r


def kruskal(n: int, xs: List[Edge]) -> List[Edge]:
  xs = sorted(xs, key=lambda e: e[2])
  txs = []
  u = UnionFind(n)
  for e in xs:
    if u.connect(e[0], e[1]):
      txs.append(e)
      if len(txs) == n-1:
        break
  return txs
